src/pymor/vectorarrays/block.py

```diff
@@ -21,11 +21,24 @@ class BlockVectorArray(VectorArrayInterface):
     :class:`~pymor.operators.block.BlockOperator`.
     """
 
+    _dtype_invalid = True # force setting dtype on first access
+
+    @invalidates_dtype
```

Comment:
The decorate feels a bit overengineered. You could simple write `self._dtype_invalid = True`. Same amount of lines of code and easier to step through with a debugger. (Also, it's not needed for `__init__`.)

Reply:
I like the decorator because it's more expressive, you can can still see which operations might modify dtype with member functions when folded and it's harder to loose in refactoring and different code paths.